### DISCUSSION TITLE
Plot IPF from ctf file with isovalue of a variable on top

### DISCUSSION THREAD

**Discussion by RBstudentPhD** [mtex-toolbox/mtex#1359] [Ask Anything] [answered]
Hi,

I'm a PhD student and a new user of MTEX. First of all, I find MTEX extraordinary. All the possibilities it offers are great. Thank you for your work and for sharing it.

The reason for my message is that I have EBSD map stored in a ctf file, and I'm trying to plot the IPF with the isovalue of the Schmid's Factor (or Young's modulus) on top of it.

For the first part of my question I do as follow:
```
fname = 'CTF\ctf_file_name.ctf';
ebsd = loadEBSD_ctf(fname,'convertSpatial2EulerReferenceFrame');
figure; plot(ebsd('Ti-Hex'),ebsd('Ti-Hex').orientations)
ori = ebsd('Ti-Hex').orientations;
odf = calcDensity(ori);
figure; plotIPDF(odf,xvector)
mtexColorbar
```
But the results I get from this are kind of funny. The IPF is not really in accordance with the EBSD map.
![EBSD](https://user-images.githubusercontent.com/100238542/155259398-1b12fea7-7e57-4135-9a79-1441ccb58883.png)
![IPF](https://user-images.githubusercontent.com/100238542/155259403-f25c89c3-3e62-4dff-9688-ee846a60f0ed.png)

For the second part, well I totally lost. I have defined my slip system:
```
cs = crystalSymmetry('622',[2.95,2.95,4.68],'x||a','mineral','Titanium (Alpha)');
sSBasal = slipSystem.basal(cs);
sSBasSym = sSBas.symmetrise('antipodal')
sSPri = slipSystem.prismaticA(cs);
sSPriSym = sSPri.symmetrise('antipodal')
sSPyr = slipSystem.pyramidalA(cs);
sSPyrSym = sSPyr.symmetrise('antipodal')
```
But I don't know how to plot the isovalue on an IPF.
In the end, I would like to have something like the following picture but instead of dots, it should be the IPF of the ODF.
![Unit-triangle-plots-of-grains-which-exhibited-a-basal-b-prismatic-c-pyramidal_W640](https://user-images.githubusercontent.com/100238542/155259870-e68c260a-e97a-4975-8448-d5c3aba65c9b.jpg)

I thank you a lot in advance for your help and also for your software.
Best regards,
RB

**Comment by kilir** [comment]
Hi,
if you think you ipf map is not correct, you probably use a different convention than you use elsewhere. Have a look here: https://mtex-toolbox.github.io/EBSDReferenceFrame.html

For the interpolation of properties on a pole figure, have a look at the forum e.g. https://github.com/mtex-toolbox/mtex/discussions/1339#discussioncomment-2132491

Cheers,
Rüdiger

**Reply by RBstudentPhD** [reply]
Hi Rüdiger,

First I would like to thank you for your support.

So I've tried the following code :

```
fname = 'CTF\ctf_file_name.ctf';
ebsd = loadEBSD_ctf(fname,'convertSpatial2EulerReferenceFrame');
ori = ebsd('Ti-Hex').orientations;
odf = calcDensity(ori); 
sSBas = slipSystem.basal(ebsd.CS);
sSBasSym = sSBas.symmetrise('antipodal');
sSBasSymLoc = ori * sSBasSym;
direction = normalize(vector3d(1,0,0));
sigma = stressTensor.uniaxial(direction);
SF = max(abs(sSBasSymLoc.SchmidFactor(sigma)),[],2);
figure; plotIPDF(ori,SF,direction)
figure; plotIPDF(ori,SF,direction,'contourf')
```

The first figure gives me the expected scatter, but when I try to plot the isovalues with the option 'contour' or 'contourf' I get the following error message:
```
Error using reshape
Number of elements must not change. Use [] as one of the size inputs to automatically calculate the appropriate size for
that dimension.

Error in S2FunHarmonic.approximation (line 38)
y = reshape(yy, [length(nodes) s(2:end)]);

Error in vector3d/interp (line 55)
    sF = S2FunHarmonic.approximation(v, y, varargin{:});

Error in vector3d/smooth (line 51)
      cdata = interp(v,cdata,S2G,'cutOutside',varargin{:});

Error in Miller/smooth (line 25)
[varargout{1:nargout}] = smooth@vector3d(m,varargin{:},sR,m.CS);

Error in vector3d/contourf (line 31)
  [varargout{1:nargout}] = v.smooth(data,'contours',10,'LineStyle','-','LineColor','k',varargin{:});

Error in vector3d/plot (line 111)
    [varargout{1:nargout}] = v.contourf(varargin{:});

Error in vector3d/subsref (line 20)
    [varargout{1:nargout}] = builtin('subsref',v,s);

Error in orientation/plotIPDF (line 82)
  [~,cax] = h.plot(repmat(data,1,length(rSym)),'symmetrised',...
```

Could please help me, and tell me what I did wrong there?

Also, bonus question, what is the difference between these two figures ? :
figure; plotIPDF(odf,direction); mtexColorbar
figure; plotIPDF(ori,direction,'contourf'); mtexColorbar

Thank you in advance,
regards,
RB

**Reply by kilir** [reply]
Hi,
this should work (using latest Metx and Matlab 2021b)
```
mtexdata titanium
ori = ebsd('Ti').orientations;

sSBas = slipSystem.basal(ebsd('Ti').CS);
sSBasSym = sSBas.symmetrise('antipodal');
sSBasSymLoc = ori * sSBasSym;

direction = normalize(vector3d(1,0,0));
sigma = stressTensor.uniaxial(direction);
SF = max(abs(sSBasSymLoc.SchmidFactor(sigma)),[],2);
figure; plotIPDF(ori,SF,direction)

figure; plotIPDF(ori,SF,direction,'contourf','bandwidth',32)
% maybe just one orientation per grain here
```

The difference between
```
plotIPDF(odf,direction); 
plotIPDF(ori,direction,'contourf'); 
```
is, that the fist plots continuous values, the latter is an iso-value plot, contoured at certain values.
Cheers,
Rüdiger

**Reply by RBstudentPhD** [reply]
Hi,

Thank you for your quick answer.
I've tried your code and I still get the same error message.
I don't really see where is the difference between your code and my code except for ",'bandwidth',32".

I'm using Matlab2021a and MTEX5.7.0 (unfortunately, the version 5.8.0 was giving an error message "Warning: missing mex-file S1Grid_find.mexw64" when using the function inpolygon). 

Regards,
RB

**Reply by RBstudentPhD** [reply]
Looking in the web, I found a way to plot what I want:
(source https://groups.google.com/g/mtexmail/c/OXG-sjFRyRU)
```
figure

cs = crystalSymmetry('6/mmm');
odf = unimodalODF(orientation.rand(1,cs));  

plotIPDF(odf,xvector,'antipodal')  
hold on  

sS = slipSystem.basal(cs);  
sS = sS.symmetrise;  
r = plotS2Grid(fundamentalSector(cs,90*degree,'antipodal'),'resolution',0.5*degree);  
tau = SchmidFactor(sS,r);  
[tauMax,id] = max(abs(tau),[],2);  

mc=mtexColorbar;  
mc_lim=mc.Limits(2);% get maximum of colorbar

contour(r,tauMax,'contour',0.05:0.05:0.5,'lineColor','k', 'linestyle',':','linewidth',2,'ShowText','on')

hold off  

CLim(gcm,[0 mc_lim]);% define colorbar range based on the IPF levels
```
The issue is that even though there is the option 'ShowText','on', well the text is not showing.
Any idea ?
![newfile](https://user-images.githubusercontent.com/100238542/155862892-e68058e6-93d6-4dcb-82aa-69e520d2f2bb.jpg)

Regards,
RB

**Reply by kilir** [reply]
Hi,
this seems to be a limitation of Matlabs clabel() refusing to label contours on small plots (e.g. if you try '321') and there's no Mtex-specific implmentation which corrects for that. Potential workaround could be to prevent clabel() to check if there's sufficient space or modify the function such that it also works on very small plots.
Cheers,
Rüdiger

**Reply by RBstudentPhD** [reply]
Apparently when i add the line "setMTEXpref('bAxisDirection','north');" right after "figure", the texts show but now the triangle is flipped ..
![newfile2](https://user-images.githubusercontent.com/100238542/156007974-e0a765d2-6345-477e-bb33-5601f53aba52.jpg)
.
Any idea on how to unflip it?

Regards,
RB

**Comment by RBstudentPhD** [comment, marked as answer]
Ok, so I've found a solution to display the value of the contour line. The code is not pretty, but the figure is :)

Here is my solution if someone is wondering how I did it:

```
figure
cs = crystalSymmetry('6/mmm');
odf = unimodalODF(orientation.rand(100,cs));  

plotIPDF(odf,xvector,'antipodal')  
hold on  

sS = slipSystem.basal(cs);  
sS = sS.symmetrise;  
r = plotS2Grid(fundamentalSector(cs,90*degree,'antipodal'),'resolution',0.5*degree);  
tau = SchmidFactor(sS,r);  
[tauMax,id] = max(abs(tau),[],2);  

mc=mtexColorbar;  
mc_lim=mc.Limits(2);% get maximum of colorbar

cm = contour(r,tauMax,'contour',[0.3 0.4 0.45 0.49],'lineColor','k', 'linestyle',':','linewidth',2,'ShowText','on');
cm = cm.ContourMatrix;

i=1; %figure; hold on
while (i<size(cm,2))
        HA1='center'; HA2='center'; VA1='top'; VA2='top'; 
      if (-cm(1,1+i))>=0.0001; VA1='bottom'; HA1='right';
        if (cm(2,1+i))>=1.2; HA1='left'; end; end
      if (-cm(1,cm(2,i)+i))>=0.0001; VA2='bottom'; HA2='right';
        if (cm(2,cm(2,i)+i))>=1.2; HA2='left'; end; end
      if (cm(2,1+i))<=1.355
      text(cm(2,1+i),-cm(1,1+i),num2str(cm(1,i)), ...
        'HorizontalAlignment',HA1,'VerticalAlignment',VA1, ...
        'FontSize',13); end
      if (cm(2,cm(2,i)+i))<=1.355
      text(cm(2,cm(2,i)+i),-cm(1,cm(2,i)+i),num2str(cm(1,i)), ...
        'HorizontalAlignment',HA2,'VerticalAlignment',VA2, ...
        'FontSize',13); end
      i=i+cm(2,i)+1;
end

hold off  
CLim(gcm,[0 mc_lim]);% define colorbar range based on the IPF levels
```
![newfile3](https://user-images.githubusercontent.com/100238542/156435944-87a811b8-fda0-4a5b-a877-94cd40fdcb6c.jpg)
